modules/shared/src/main/scala/org/tessellation/ext/swaydb/SwaydbRef.scala

```diff
@@ -0,0 +1,157 @@
+                      }
+                    }
+                  case Some(newV) =>
+                    guarded {
+                      delay {
+                        hasBeenCalled.compareAndSet(false, true) && db.put(k, newV).isInstanceOf[OK]
+                      }
+                    }
+                }
+              }
+              (None, set)
+            case Some(init) =>
+              val set: Option[V] => F[Boolean] = { (opt: Option[V]) =>
+                opt match {
+                  case None =>
+                    guarded {
+                      delay {
+                        hasBeenCalled.compareAndSet(false, true) && db.remove(k).isInstanceOf[OK]
+                      }
+                    }
+                  case Some(newV) =>
+                    guarded {
+                      delay {
+                        hasBeenCalled.compareAndSet(false, true) && db.update(k, newV).isInstanceOf[OK]
+                      }
+                    }
+                }
+              }
+              (Some(init), set)
+          }
+        }
+
+      def get: F[Option[V]] =
+        delay {
+          db.get(k)
+        }
+
+      def modify[B](f: Option[V] => (Option[V], B)): F[B] = {
+        lazy val loop: F[B] = tryModify(f).flatMap {
+          case None    => loop
+          case Some(b) => concurrent.pure(b)
+        }
+        loop
+      }
+
+      def modifyState[B](state: State[Option[V], B]): F[B] =
+        modify(state.run(_).value)
+
+      def set(a: Option[V]): F[Unit] =
+        a match {
+          case None    => guarded { delay { db.remove(k); () } }
+          case Some(v) => guarded { delay { db.put(k, v); () } }
+        }
+
+      def tryModify[B](f: Option[V] => (Option[V], B)): F[Option[B]] =
+        guarded {
+          delay {
+            db.get(k) match {
+              case None =>
+                f(None) match {
+                  case (None, b) =>
+                    // no-op
+                    concurrent.pure(b.some)
+                  case (Some(newV), b) =>
+                    delay { db.put(k, newV) }.map(_ => b.some)
+                }
+              case Some(init) =>
+                f(Some(init)) match {
+                  case (None, b) =>
+                    delay { db.remove(k) }.map(_ => b.some)
+                  case (Some(next), b) =>
+                    delay { db.update(k, next) }.map(_ => b.some)
+                }
+            }
+          }.flatten
+        }
+
+      def tryModifyState[B](state: State[Option[V], B]): F[Option[B]] =
+        tryModify(state.run(_).value)
+
+      def tryUpdate(f: Option[V] => Option[V]): F[Boolean] =
+        tryModify { opt =>
+          (f(opt), ())
+        }.map(_.isDefined)
+
+      def update(f: Option[V] => Option[V]): F[Unit] = {
+        lazy val loop: F[Unit] = tryUpdate(f).flatMap {
+          case true  => concurrent.unit
+          case false => loop
+        }
```

Comment:
Alternatively: `tryUpdate(f).ifM(concurrent.unit, loop)`

Reply:
Implementation is copied from cats